aws/_base/main.tf

```diff
@@ -12,30 +12,42 @@ provider "aws" {
   region  = "us-east-1"
 }
 
-resource "aws_spot_instance_request" "runner" {
-  ami                  = var.ami
-  instance_type        = var.instance_type
-  key_name             = "gitlab-runner"
-  wait_for_fulfillment = true
-  spot_type            = "one-time"
+resource "aws_spot_fleet_request" "runner" {
+  allocation_strategy = "lowestPrice"
 
-  subnet_id = var.internal_network ? data.aws_subnet.internal_subnet_primary.id : data.aws_subnet.external_subnet_primary.id
+  fleet_type                          = "request"
+  iam_fleet_role                      = local.iam_fleet_role
+  target_capacity                     = 1
+  wait_for_fulfillment                = "true"
+  terminate_instances_with_expiration = "true"
 
-  vpc_security_group_ids = [
-    var.internal_network ? data.aws_security_group.internal_security_group.id : data.aws_security_group.external_security_group.id
-  ]
+  dynamic "launch_specification" {
+    for_each = setproduct(var.instance_types, var.internal_network ? local.internal_subnets : local.external_subnets)
+    content {
+      ami           = var.ami
+      subnet_id     = launch_specification.value[1]
+      key_name      = "gitlab-runner"
+      instance_type = launch_specification.value[0]
```

Comment:
The weird `for_each` syntax is just for dynamically generating all possible combinations for the actual spot fleet requests. It's AWS who decides what's launched, not terraform. Terraform is just listing the options. I screenshotted the AWS UI to demonstrate this:

![image](https://user-images.githubusercontent.com/119861/170053070-1e9755cc-ffd1-4074-87d3-16be848465af.png)

With our current setting, AWS will pick the **cheapest** (allocation strategy) and **available** instance type in a specific zone. We set which instance types and which zones can be used. Does it make sense?

Reply:
For this particular request, AWS chose `c6a.large` and `subnet-ddc07282` because it was apparently the cheapest available option at that time. Magic.

![image](https://user-images.githubusercontent.com/119861/170053624-d6579a5d-379f-4733-839d-b7b10d6b75a6.png)

Reply:
Oh wow that's really cool, thanks for the explanation and demonstration!